students/andykwon/Lesson09/mailroom_pt5.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python3
+
+"""
+simple mailroom program for a non profit
+"""
+import sys
+
+
+class Donations():
+
+    def __init__(self, donations=None):
+
+        if donations is None:
+            self.donations = {}
+        else:
+            self.donations = donations
+
+    def __str__(self):
+        return str(self.donations)
+
+    def add_donation(self, name, amount):
+        self.donations.setdefault(name, []).append(amount)
+
+    def generate_report(self):
+        report = []
+        report_data = []
+
+        header = ("{:<25s} | {:^15s} | {:^10s} | {:15s}".format(
+            "Donor Name", "Total Given", "Num Gifts", "Average Gift"))
+        divider = "_" * len(header)
+
+        report.append(header)
+        report.append(divider)
+
+        for key_name in self.donations:
+            name = key_name
+            total_given = sum(self.donations[key_name])
+            total_count = len(self.donations[key_name])
+            avg_given = total_given / total_count
+            report_data.append((name, total_given, total_count, avg_given))
+
+        # Sort with respect to "Total Given" value.
+        report_data.sort(key=donations, reverse=True)
+
+        # Adding the data into report with formatting
+        for data in report_data:
+            report.append(
+                "{:25s}   ${:14.2f}   {:10d}   ${:11.2f}".format(* data))
+
+        printable_report = ("\n").join(report)
+
+        return printable_report
+
+    def generate_letter(self):
+
+        print("== Send letters to everyone ==")
+
+        for key_name in self.donations:
+            file = open("Letter_for_" + key_name + ".txt", "w")
+
+            letter = ("Dear " + key_name + ",\n\n" +
+                      "Thank you for your recent donation of $" +
+                      "%.2f" % self.donations[key_name][-1] + ".\n\n" +
+                      "It will be put to very good use.\n\n" +
+                      "Sincerely,\n\n" + "-The Team")
+
+            file.write(letter)
+
+            file.close()
+
+    # @property
+    # def donation_count(self, name):
+    #     len(self.donations[name])
+
+    # @property
+    # def donation_total(self, name):
+    #     sum(self.donations[name])
+
+    # @property
+    # def donation_average(self, name):
+    #     sum(self.donations[name]) / len(self.donations[name])
+
+    @property
+    def last_donated(self, name):
+        return self.donations[name][-1]
+
+    def names_of_donors(self):
+
+        donors = []
+
+        donors.append("Here is the list of donors:")
+
+        for name in self.donations:
+            donors.append(name)
+
+        printable_donors = "\n".join(donors)
+
+        return printable_donors
+
+
+# -------------------------------------------------------------
```

Comment:
Ideally we wanted two classes -- one with donor-specific data, the other with the donor database, made up of of donors, but this works